File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* l1 = head;
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }

        ListNode* secondListStarts = slow->next;
        // severing connection in middle 
        ListNode* prev = nullptr;
        ListNode* curr = secondListStarts;
        slow -> next = nullptr;
        // Reversing second half now
        while(curr){
            ListNode* nex = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nex;
        }
        ListNode* l2 = prev;
        
        // Now take one node from each
        while(l2){
            // take from first list
            ListNode* tmp1 = l1->next;
            ListNode* tmp2 = l2->next;

            l1->next = l2;
            l1 = tmp1;
            l2->next = l1;
            l2 = tmp2;
        }
        
    }
};

File: Data Structures & Algorithms/n-ary-tree-postorder-traversal/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: Optional[int] = None, children: Optional[List['Node']] = None):
        self.val = val
        self.children = children
"""

class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        res = []

        def postorder(curr):
            if curr == None:
                return

            for c in curr.children:
                postorder(c)
            res.append(curr.val)

        postorder(root)
        return res
        